core/src/main/scala/com/salesforce/op/stages/impl/insights/RecordInsightsLOCO.scala

```diff
@@ -93,107 +91,179 @@ class RecordInsightsLOCO[T <: Model[T]]
-    baseScore: Array[Double]): Array[Double] = {
-    featureArray.update(i, (oldInd, 0))
-    val score = modelApply(labelDummy, OPVector(Vectors.sparse(featureSize, featureArray))).score
+  private def computeDiffs
+  (
+    i: Int,
+    oldInd: Int,
+    oldVal: Double,
+    featureArray: Array[(Int, Double)],
+    featureSize: Int,
+    baseScore: Array[Double]
+  ): Array[Double] = {
+    featureArray.update(i, (oldInd, 0.0))
+    val score = modelApply(labelDummy, Vectors.sparse(featureSize, featureArray).toOPVector).score
     val diffs = baseScore.zip(score).map { case (b, s) => b - s }
     featureArray.update(i, (oldInd, oldVal))
     diffs
   }
 
-  private def returnTopPosNeg(filledSize: Int, featureArray: Array[(Int, Double)], featureSize: Int,
-    baseScore: Array[Double], k: Int, indexToExamine: Int): Seq[(Int, Double, Array[Double])] = {
-    // Heap that will contain the top K positive LOCO values
-    val positiveMaxHeap = PriorityQueue.empty(MinScore)
-    // Heap that will contain the top K negative LOCO values
-    val negativeMaxHeap = PriorityQueue.empty(MaxScore)
-    // for each element of the feature vector != 0.0
-    // Size of positive heap
-    var positiveCount = 0
-    // Size of negative heap
-    var negativeCount = 0
-    var i = 0
-    while (i < filledSize) {
+  private def sumArrays(left: Array[Double], right: Array[Double]): Array[Double] = {
+    left.zipAll(right, 0.0, 0.0).map { case (l, r) => l + r }
+  }
+
+  private def returnTopPosNeg
+  (
+    featureArray: Array[(Int, Double)],
+    featureSize: Int,
+    baseScore: Array[Double],
+    k: Int,
+    indexToExamine: Int
+  ): Seq[LOCOValue] = {
+
+    val minMaxHeap = new MinMaxHeap(k)
+    val aggregationMap = mutable.Map.empty[String, (Array[Int], Array[Double])]
+    for {i <- featureArray.indices} {
       val (oldInd, oldVal) = featureArray(i)
       val diffToExamine = computeDiffs(i, oldInd, oldVal, featureArray, featureSize, baseScore)
-      val max = diffToExamine(indexToExamine)
+      val history = histories(oldInd)
 
-      if (max > 0.0) { // if positive LOCO then add it to positive heap
-        positiveMaxHeap.enqueue((i, max, diffToExamine))
-        positiveCount += 1
-        if (positiveCount > k) { // remove the lowest element if the heap size goes from 5 to 6
-          positiveMaxHeap.dequeue()
+      // Let's check the indicator value and descriptor value
+      // If those values are empty, the field is likely to be a derived text feature (hashing tf output)
+      if (history.indicatorValue.isEmpty && history.descriptorValue.isEmpty) {
+        // Name of the field
+        val rawName = history.parentFeatureStages match {
+          case s if s.exists(_.contains(smartTextClassName)) => history.parentFeatureOrigins.headOption
+          case s if s.exists(_.contains(smartTextMapClassName)) => history.grouping
+          case _ => None
```

Comment:
If the derived text feature is not from a SmartTextVectorizer then it is added neither to `minMaxHeap` nor `aggregationMap`. Basically its LOCO value is being ignored. I think that's incorrect we should add it to `minMaxHeap`.   @mweilsalesforce What do you think ?

Reply:
I would prefer to completely remove dependency of specific stage class and only decide based on the grouping & indicator values. is that even possible? (cc @Jauntbox )

Reply:
I don't know the capabilities of TMOG regard Text engineering, but do you mean @sanmitra that any text derived feature with no indicator/descriptor value won't be able to be interpreted by LOCO (from other stages like OPLDA, OPWord2Vec, ...) ?

Reply:
yes.

Reply:
Ok. The rule is now : "If a derived text feature has no indicator/descriptor value, then it will be aggregated. Aggregation by `parentFeatureName + parentFeatureStage`


Let me put the changes

Reply:
@mweilsalesforce are you planning to do it in this PR?

Reply:
Done